sympy/matrices/expressions/trace.py

```diff
@@ -144,8 +145,11 @@ def get_arg_key(x):
 
     def _eval_rewrite_as_Sum(self, expr, **kwargs):
         from sympy.concrete.summations import Sum
-        i = Dummy('i')
-        return Sum(self.arg[i, i], (i, 0, self.arg.rows-1)).doit()
+        i = uniquely_named_symbol('i', expr)
+        s = Sum(self.arg[i, i], (i, 0, self.arg.rows - 1))
+        if not isinstance(expr, MatrixExpr):
+            s = s.doit()
```

Comment:
If this call to `doit` would otherwise raise `IndexError` then that suggests that something should still be fixed somewhere because `doit` shouldn't raise `IndexError`.

Reply:
Have a look at the `eval_sum_hyper` routine. It needs to access the `upper bound + 1`: `res2 = _eval_sum_hyper(f, i, b + 1)`. This will fail for any indexed object which cannot be computed at that bound. I left a note on the routine regarding indexed objects and I added a safeguard in the trace routine that is writing trace of a matrix as a sum. In the case of an actual matrix the dimensions are concrete and `eval_sum_hyper` is not called.

Reply:
Can you show example code needed to set this up so I can see the error? Would I see the error on master or only with the changes in this PR?

Reply:
Saying that it shouldn't happen is equivalent to saying either "do routine writing sums with indexed objects should call the `eval_sum_hyper` routine. But calling that routine is part of the `Sum.doit` machinery. So any routine writing Sum with indexed objects should not call doit. And that is the change that I made in `trace`.

Reply:
What I was looking for was how to reproduce this. I see now that with this PR it's:
```python
In [1]: n = Symbol('n', integer=True)

In [2]: M = MatrixSymbol('M', n, n)

In [3]: trace(M).rewrite(Sum).doit()
---------------------------------------------------------------------------
IndexError 
```
Strangely the `integer=True` is needed here.

As you say it goes wrong from here:
https://github.com/sympy/sympy/blob/14cf005b8fea132207654603cc6495900040106a/sympy/concrete/summations.py#L1310-L1311
There could be lots of situations where just substituting `b+1` in might break something so either the exception needs to be caught here (i.e. your previous code) or there should be some code to determine whether this is an appropriate function to use in this routine at all. There isn't much documentation but I presume that `eval_sum_hyper` is only supposed to work for hypergeometric functions and `M[i, i]` is clearly not hypergeometric in `i`.

Reply:
Catching the exception is not exactly satisfactory because this kind of potentially invalid operation could lead to all sorts of different exceptions.

Reply:
These lines seem to be responsible for this output for example:
```python
In [1]: b = symbols('b', integer=True, positive=True)

In [2]: summation(1/(x-b)**2, (x, 0, b-1))
Out[2]: 
 ⎛         ⅈ⋅π⎞      
Φ⎝1, 2, b⋅ℯ   ⎠ + zoo
```

Reply:
This is giving zoo in cases where actually the result might be well defined:
```python
In [7]: Sum(1/(x-b)**2, (x, 0, b-1)).subs(b, 2).doit()
Out[7]: 5/4

In [8]: b = symbols('b', integer=True, positive=True)

In [9]: s = Sum(1/(x-b)**2, (x, 0, b-1))

In [10]: s.subs(b, 2).doit()
Out[10]: 5/4

In [11]: s.doit().subs(b, 2)
Out[11]:
---------------------------------------------------------------------------
ZeroDivisionError  
```

Reply:
I added a `is_hypergeometric` tests to the routine and then a check that illegals are not introduced by the `b + 1` evaluation. From the wolframalpha result, it appears that a solution is possible. That's a separate issue; at least we can check that there is no `zoo`.

Reply:
I'm not sure how well `is_hypergeometric` works. Note that it just calls `hypersimp` which is already being called here I think.

Reply:
bad evaluations are avoid now and tested as:
```python
def test_pr_22677():
    b = Symbol('b', integer=True, positive=True)
    assert Sum(1/x**2,(x, 0, b)).doit() == Sum(x**(-2), (x, 0, b))
    assert Sum(1/(x - b)**2,(x, 0, b-1)).doit() == Sum(
        (-b + x)**(-2), (x, 0, b - 1))
```
The first one gives the following in master:
```python
>>> Sum(1/x**2,(x,0,b)).doit()
zoo - (b**2 + 2*b + 1)*lerchphi(1, 2, b + 1)/(b + 1)**2
```